kafka/consumer_test.go

```diff
@@ -18,12 +18,238 @@
+				SASL:      saslplain.New(saslplain.Plain{}),
+				TLS:       &tls.Config{},
+			},
+			expectErr: false,
+		},
+	}
+	for name, tc := range testCases {
+		t.Run(name, func(t *testing.T) {
+			consumer, err := NewConsumer(tc.cfg)
+			if err == nil {
+				defer assert.NoError(t, consumer.Close())
+			}
+			if tc.expectErr {
+				assert.Error(t, err)
+				assert.Nil(t, consumer)
+			} else {
+				assert.NoError(t, err)
+				assert.NotNil(t, consumer)
+			}
+		})
+	}
+}
+
+func TestConsumerHealth(t *testing.T) {
+	testCases := map[string]struct {
+		expectErr  bool
+		closeEarly bool
+	}{
+		"success": {
+			expectErr:  false,
+			closeEarly: false,
+		},
+		"failure": {
+			expectErr:  true,
+			closeEarly: true,
+		},
+	}
+	for name, tc := range testCases {
+		t.Run(name, func(t *testing.T) {
+			cluster, err := kfake.NewCluster()
+			require.NoError(t, err)
+
+			addrs := cluster.ListenAddrs()
+			consumer := newConsumer(t, ConsumerConfig{
+				Brokers:   addrs,
+				Topics:    []string{"topic"},
+				GroupID:   "groupid",
+				Decoder:   json.JSON{},
+				Logger:    zap.NewNop(),
+				Processor: model.ProcessBatchFunc(func(context.Context, *model.Batch) error { return nil }),
+			})
+
+			if tc.closeEarly {
+				cluster.Close()
+			} else {
+				defer cluster.Close()
+			}
+
+			if tc.expectErr {
+				assert.Error(t, consumer.Healthy())
+			} else {
+				assert.NoError(t, consumer.Healthy())
+			}
+		})
+	}
+}
+
+func TestConsumerFetch(t *testing.T) {
+	event := model.APMEvent{Transaction: &model.Transaction{ID: "1"}}
+	topics := []string{"topic"}
+	client, addrs := newClusterWithTopics(t, topics...)
+
+	cfg := ConsumerConfig{
+		Brokers: addrs,
+		Topics:  topics,
+		GroupID: "groupid",
+		Decoder: json.JSON{},
+		Logger:  zap.NewNop(),
+		Processor: model.ProcessBatchFunc(func(ctx context.Context, b *model.Batch) error {
+			assert.Len(t, *b, 1)
+			assert.Equal(t, event, (*b)[0])
+			return nil
+		}),
+	}
+
+	b, err := json.JSON{}.Encode(event)
+	require.NoError(t, err)
+
+	record := &kgo.Record{
+		Topic: topics[0],
+		Value: b,
+	}
+
+	results := client.ProduceSync(context.Background(), record)
+	assert.NoError(t, results.FirstErr())
+
+	r, err := results.First()
+	assert.NoError(t, err)
+	assert.NotNil(t, r)
+
+	consumer := newConsumer(t, cfg)
+	assert.NoError(t, consumer.fetch(context.Background()))
+}
+
+func TestMultipleConsumers(t *testing.T) {
+	event := model.APMEvent{Transaction: &model.Transaction{ID: "1"}}
+	topics := []string{"topic"}
+	client, addrs := newClusterWithTopics(t, topics...)
+
+	var count atomic.Int32
+	cfg := ConsumerConfig{
+		Brokers: addrs,
+		Topics:  topics,
+		GroupID: "groupid",
+		Decoder: json.JSON{},
+		Logger:  zap.NewNop(),
+		Processor: model.ProcessBatchFunc(func(ctx context.Context, b *model.Batch) error {
+			count.Add(1)
+			assert.Len(t, *b, 1)
+			//assert.Equal(t, event, (*b)[0])
+			return nil
+		}),
+	}
+
+	b, err := json.JSON{}.Encode(event)
```

Comment:
Let's use the same codec for producing and consuming. That'll make it easier to swap in the future too without changing multiple lines.

```suggestion
	codec := json.JSON{}

	cfg := ConsumerConfig{
		Brokers: addrs,
		Topics:  topics,
		GroupID: "groupid",
		Decoder: codec,
		Logger:  zap.NewNop(),
		Processor: model.ProcessBatchFunc(func(_ context.Context, b *model.Batch) error {
			assert.Len(t, *b, 1)
			assert.Equal(t, event, (*b)[0])
			return nil
		}),
	}

	b, err := codec.Encode(event)
```